### ISSUE TITLE
Mongo error when querying project details

### ISSUE THREAD

**Issue by anishkargaonkar** [sorry-cypress/sorry-cypress#319]
## Summary
Not able to query project details. Failing after running 10-12 spec files (total 40).
Error:  Mongo error `BSONObjectTooLarge` or `Location4568`

Upon investigating, looks like this could be caused because of `$lookup` query in getRun API.

curl
```
curl '<domain>/api' \
  -H 'Connection: keep-alive' \
  -H 'Pragma: no-cache' \
  -H 'Cache-Control: no-cache' \
  -H 'accept: */*' \
  -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.128 Safari/537.36' \
  -H 'content-type: application/json' \
  -H 'Origin: <domain>' \
  -H 'Referer: <domain>' \
  -H 'Accept-Language: en-GB,en-US;q=0.9,en;q=0.8' \
  --data-raw $'{"operationName":"getRunSummary","variables":{"runId":"61c80bc25c0f6a20e90fdca712bdac9d"},"query":"query getRunSummary($runId: ID\u0021) {\\n  run(id: $runId) {\\n    runId\\n    createdAt\\n    meta {\\n      ...RunSummaryMeta\\n      __typename\\n    }\\n    completion {\\n      ...RunSummaryCompletion\\n      __typename\\n    }\\n    specs {\\n      ...RunSummarySpec\\n      __typename\\n    }\\n    __typename\\n  }\\n}\\n\\nfragment RunSummaryMeta on RunMeta {\\n  ciBuildId\\n  projectId\\n  commit {\\n    sha\\n    branch\\n    remoteOrigin\\n    message\\n    authorEmail\\n    authorName\\n    __typename\\n  }\\n  __typename\\n}\\n\\nfragment RunSummaryCompletion on RunCompletion {\\n  completed\\n  inactivityTimeoutMs\\n  __typename\\n}\\n\\nfragment RunSummarySpec on FullRunSpec {\\n  claimed\\n  results {\\n    stats {\\n      ...AllInstanceStats\\n      __typename\\n    }\\n    __typename\\n  }\\n  __typename\\n}\\n\\nfragment AllInstanceStats on InstanceStats {\\n  suites\\n  tests\\n  pending\\n  passes\\n  failures\\n  skipped\\n  suites\\n  wallClockDuration\\n  wallClockStartedAt\\n  wallClockEndedAt\\n  __typename\\n}\\n"}' \
  --compressed \
  --insecure
```

response:
```
{
  "errors": [
    {
      "message": "Total size of documents in instances matching pipeline's $lookup stage exceeds 104857600 bytes",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "run"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": {
          "ok": 0,
          "code": 4568,
          "codeName": "Location4568",
          "name": "MongoError",
          "stacktrace": [
            "MongoError: Total size of documents in instances matching pipeline's $lookup stage exceeds 104857600 bytes",
            "    at MessageStream.messageHandler (/app/node_modules/mongodb/lib/cmap/connection.js:268:20)",
            "    at MessageStream.emit (events.js:315:20)",
            "    at processIncomingData (/app/node_modules/mongodb/lib/cmap/message_stream.js:144:12)",
            "    at MessageStream._write (/app/node_modules/mongodb/lib/cmap/message_stream.js:42:5)",
            "    at writeOrBuffer (internal/streams/writable.js:358:12)",
            "    at MessageStream.Writable.write (internal/streams/writable.js:303:10)",
            "    at Socket.ondata (internal/streams/readable.js:719:22)",
            "    at Socket.emit (events.js:315:20)",
            "    at addChunk (internal/streams/readable.js:309:12)",
            "    at readableAddChunk (internal/streams/readable.js:284:9)"
          ]
        }
      }
    }
  ],
  "data": {
    "run": null
  }
}

```

Sorry cypress shows **No runs found** but should show the run details.

![image](https://user-images.githubusercontent.com/32673168/115249703-810fa180-a146-11eb-83f1-2ccf70806723.png)
![image](https://user-images.githubusercontent.com/32673168/115250243-03986100-a147-11eb-8f67-6f4cc9d6863d.png)

## How to reproduce
- Use mongo driver
- Run cypress in parallel with around 40 spec files (it fails after running 10-12 spec files)

## Environment

- sorry-cypress version: `v1.0.0-rc.5`
- cypress version: `7.1.0`
- platform: `azure` [helm chart](https://github.com/sorry-cypress/charts)
- service: `api | dashboard`

**Comment by anishkargaonkar** [comment]
@agoldis I can pick this up. Can you point me in some direction to get started?

**Comment by agoldis** [comment]
@anishkargaonkar Sure! Start with https://docs.sorry-cypress.dev/development/development-guide
and the problematic query is [here](https://github.com/sorry-cypress/sorry-cypress/blob/635d71750f9453e5193534938fb77e36416db120/packages/api/src/datasources/runs.ts#L143)

One solution might be using GraphQL fields resolver  (or explicit query) to avoid using lookups.

**Comment by agoldis** [comment]
@anishkargaonkar any progress on that?

**Comment by anishkargaonkar** [comment]
@agoldis I've started working on graphQL query resolver, taking some time since I've never worked with graphQL.

Schema for full specs is already there, working on adding a query [here](https://github.com/sorry-cypress/sorry-cypress/blob/5cee075e7a951e1cf11b06f453334e2d3fcef015/packages/api/src/resolvers/index.ts#L98)

**Comment by agoldis** [comment]
@anishkargaonkar awesome! please let me know if you need any help

**Comment by anishkargaonkar** [comment]
Thank you @agoldis, do let me know if there are any issues with PR #336. One thing I noticed, the query is a bit slow, for 136 specs in one run it took around ~7 seconds for the response (mongodb instance was running on a remote host). I think it'll be an issue when there are more runs in a project. Is there any way we can optimize it?

**Comment by xtroncode** [comment]
I am getting the following error with rc.9  in the getRunSummary operation : `Total size of documents in instances matching pipeline { $match: { $and: [ { instanceId: { $in: [ \"eb163273-a9c4-4c7c-af0f-1453e7da7cfe\", \"557c98d4-0..... \"5880b448-2505-4363-ae31-f5ff8b568346\" ] } }, {} ] } } exceeds maximum document size"`  does this seem related or should I file a separate issue ?

**Comment by anishkargaonkar** [comment]
@xtroncode this is the same error. Due to the aggregation query, the document size is more than mongo's allowed maximum document size (~16 MB) and that's why it's failing. 

@agoldis any update on this?

**Comment by xtroncode** [comment]
Thanks @anishkargaonkar ! Will #336 fix this as well ? or is there any other way to fix this (may be by increasing the allowed document size for now) ?

**Comment by agoldis** [comment]
Sorry @anishkargaonkar, will try to merge those today

**Comment by agoldis** [comment]
Resolved in #319
